core/job/service/job_service.go

```diff
@@ -859,34 +877,43 @@ func (j *JobService) bulkDelete(ctx context.Context, jobTenant tenant.Tenant, to
 	return deletedJobNames, me.ToErr()
 }
 
-func (*JobService) differentiateSpecs(existingJobs []*job.Job, specs []*job.Spec, jobNamesWithInvalidSpec []job.Name) (added, modified, deleted, unmodified []*job.Spec, err error) {
-	// TODO: consider checking multi-error if it is required here
-	me := errors.NewMultiError("differentiate specs errors")
-
-	var addedSpecs, modifiedSpecs, unmodifiedSpecs, deletedSpecs []*job.Spec
+func (j *JobService) differentiateSpecs(existingJobs []*job.Job, incomingSpecs []*job.Spec, jobNamesWithInvalidSpec []job.Name) (added, modified, deleted, unmodified, dirty []*job.Spec) {
+	var addedSpecs, modifiedSpecs, unmodifiedSpecs, deletedSpecs, unmodifiedDirtySpecs []*job.Spec
 
 	existingSpecsMap := job.Jobs(existingJobs).GetNameAndSpecMap()
 	for _, jobNameToSkip := range jobNamesWithInvalidSpec {
+		j.logger.Error("received invalid spec for %s", jobNameToSkip)
 		delete(existingSpecsMap, jobNameToSkip)
 	}
 
-	for _, incomingSpec := range specs {
-		if spec, ok := existingSpecsMap[incomingSpec.Name()]; !ok {
+	for _, incomingSpec := range incomingSpecs {
+		spec, ok := existingSpecsMap[incomingSpec.Name()]
+		if !ok {
 			addedSpecs = append(addedSpecs, incomingSpec)
-		} else if !reflect.DeepEqual(spec, incomingSpec) {
-			modifiedSpecs = append(modifiedSpecs, incomingSpec)
-		} else {
-			unmodifiedSpecs = append(unmodifiedSpecs, incomingSpec)
+			continue
+		}
+
+		wasDirty := spec.IsDirty()
+		spec.SetDirty(false) // need to do this for comparison(DeepEqual) with the incoming spec which is never dirty
```

Comment:
Can we do this some other way, it does not seem like a good way.

Reply:
sure, fixed it